### PR TITLE
🐛 Bug fixes and improvements related to the coordinate system

### PR DESCRIPTION
**Pull request by wlambooy**
## Description

This PR is a collection of bug fixes in order to make the experience working with layouts of different coordinate types more seamless. In particular, the following changes are introduced:

* All coordinate types now have a `wrap` function added to their respective namespaces that takes a dimension coordinate. When the coordinate is beyond the given dimension, it wraps to the next coordinate within the dimension. If no such coordinate exists, it becomes a dead copy of the dimension coordinate.
* Coordinate iteration now does not go into an infinite loop when an out of bounds iterator is given.
* Area and volume computation is now dependent on the coordinate type
* Cartesian layouts with SiQAD coordinates now always represent full dimer rows; this means that the z dimension is always 1.
* An `integral_abs` function defined on integral number types was added to `math_utils.hpp`. Simply using `std::abs` could give an ambiguity error or a compiler warning when applied to an unsigned integer. This function handles these concerns such that it can be applied to any integral number type without compiler errors or warnings.
* Addition and subtraction is now defined properly on SiQAD coordinates.
* `static_assert` checks were added to `apply_gate_layout` and `relative_to_absolute_cell_position`, making sure that the `CellLyt` template parameter is not based in SiQAD coordinates.
* Some changes to the docs: the `siqad` namespace in `coordinates.hpp` had missing docs; the "Offset coordinates are required" line was removed from `cartesian_layout.hpp`.
* In `layout_utils.hpp`, converting between SiQAD and fiction coordinates now appropriately resizes the layout. Note that since layouts with SiQAD coordinates must always have full dimer rows, the area of the transformed layout is not always equal to the area of the original layout when converting _to_ SiQAD coordinates. The tests have thus been updated accordingly.
* The `print_charge_layout` function was adapted to the new changes. Also layout cropping is now optional (default false).

Tests were added for all of the new behaviour. The `wrap` function implementations are tested through the coordinate iterator tests.

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: include/fiction/layouts/coordinates.hpp [hunk 1]
```diff
@@ -145,6 +147,39 @@ struct ucoord_t
     {
         return ucoord_t{static_cast<uint64_t>(*this) | static_cast<uint64_t>(ucoord_t{})};
     }
+    /**
+     * Clamps the coordinate to the given dimension. Lower bounds are neglected and should be handled separately if need
+     * be. When the coordinate is outside the given dimension, it is transformed to the next coordinate that is in it.
+     * If no such coordinate exists, the coordinate becomes a dead copy of the dimension.
+     *
+     * @param dimension Dimension or aspect ratio to clamp the coordinate to.
+     */
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:156]: I don't think I understand the docstring text yet. Could you maybe rephrase it and also add a quick note on the use case of this function?

File: test/io/print_layout.cpp [hunk 2]
```diff
@@ -213,98 +214,227 @@ TEST_CASE("Print empty charge layout", "[print-charge-layout]")
...
+    cl.assign_charge_state({29, 14, 0}, sidb_charge_state::NEGATIVE);
+    cl.assign_charge_state({30, 15, 0}, sidb_charge_state::NEGATIVE);
+    cl.assign_charge_state({34, 16, 0}, sidb_charge_state::POSITIVE);
+    cl.assign_charge_state({36, 17, 0}, sidb_charge_state::NEUTRAL);
+    cl.assign_charge_state({40, 18, 0}, sidb_charge_state::NEGATIVE);
+    cl.assign_charge_state({42, 19, 0}, sidb_charge_state::NEGATIVE);
+
+    SECTION("Cropped")
+    {
+        std::stringstream print_stream{};
```
Comment (marcelwa) [inline test/io/print_layout.cpp:248]: You should be able to declare the stream outside of the sections. This would prevent the duplicate.

File: test/layouts/cartesian_layout.cpp [hunk 3]
```diff
@@ -247,3 +247,11 @@ TEST_CASE("Cartesian cardinal operations", "[cartesian-layout]")
...
         });
 }
+
+TEST_CASE("Cartesian layouts with SiQAD coordinates must have a z dimension of 1")
+{
+    using lyt = cartesian_layout<siqad::coord_t>;
+
+    CHECK(lyt{aspect_ratio<lyt>{0, 0}}.z() == 1);
+    CHECK(lyt{aspect_ratio<lyt>{9, 9}}.z() == 1);
+}
```
Comment (marcelwa) [inline test/layouts/cartesian_layout.cpp:257]: Please make sure that all files end with a newline.

File: test/layouts/cartesian_layout.cpp [hunk 4]
```diff
@@ -247,3 +247,11 @@ TEST_CASE("Cartesian cardinal operations", "[cartesian-layout]")
                 std::set<cartesian_layout<offset::ucoord_t>::coordinate>{{{4, 5}, {5, 4}, {6, 5}, {5, 6}}}.count(adj));
         });
 }
+
+TEST_CASE("Cartesian layouts with SiQAD coordinates must have a z dimension of 1")
+{
+    using lyt = cartesian_layout<siqad::coord_t>;
+
+    CHECK(lyt{aspect_ratio<lyt>{0, 0}}.z() == 1);
+    CHECK(lyt{aspect_ratio<lyt>{9, 9}}.z() == 1);
```
Comment (marcelwa) [inline test/layouts/cartesian_layout.cpp:256]: Just for completeness, also add one check that initializes the layout with an aspect ratio that already has a z value of 1.

File: test/layouts/coordinates.cpp [hunk 5]
```diff
@@ -173,13 +176,100 @@ TEMPLATE_TEST_CASE("Coordinate iteration", "[coordinates]", offset::ucoord_t, cu
...
+            good_bound_coord_vector.clear();
+            good_bound_coord_vector.reserve(8);
+
+            c_lyt.foreach_coordinate(fill_coord_vector, {}, bad_bound);
+            c_lyt.foreach_coordinate(fill_good_bound_coord_vector, {}, good_bound);
+
+            if (coord_vector.size() != good_bound_coord_vector.size())
+            {
+                std::cout << bad_bound.str() << good_bound.str() << std::endl;
+            }
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:199]: Please remove debug output

File: test/layouts/coordinates.cpp [hunk 6]
```diff
@@ -173,13 +176,100 @@ TEMPLATE_TEST_CASE("Coordinate iteration", "[coordinates]", offset::ucoord_t, cu
...
+    using coord = siqad::coord_t;
+
+    CHECK(coord{-4, 4, 1} + coord{1, -7, 1} == coord{-3, -2, 0});
+    CHECK(coord{-4, 4, 1} + coord{1, -7, 0} == coord{-3, -3, 1});
+
+    CHECK(coord{-4, 4, 0} - coord{1, -7, 1} == coord{-5, 10, 1});
+    CHECK(coord{-4, 4, 1} - coord{1, -7, 1} == coord{-5, 11, 0});
+}
+
+TEST_CASE("Addition / subtraction of cube coordinates", "[coordinates]")
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:267]: This says "cube coordinates" but uses SiQAD coordinates. Which one is correct?

File: test/utils/math_utils.cpp [hunk 7]
```diff
@@ -48,3 +51,11 @@ TEST_CASE("round_to_n_decimal_places should round an input number to n decimal p
...
     }
 }
+
+TEMPLATE_TEST_CASE("abs should compute the absolute value of a number of different types", "[abs]", int8_t, int16_t,
+                   int32_t, int64_t, uint8_t, uint16_t, uint32_t, uint64_t, float, double, long double)
+{
+    auto x = static_cast<TestType>(-42);
+
+    CHECK(abs(x) == (x < 0 ? -x : x));
+}
```
Comment (marcelwa) [inline test/utils/math_utils.cpp:61]: see above

**Review by marcelwa — CHANGES_REQUESTED** [review]
Many thanks for these great additions and fixes. Looks very good overall. You have really improved the code stability! Just a couple of minor comments from my side here.

**Comment by wlambooy** [comment]
It seems that some compilers generate a warning when the result of the new abs function is casted to an integer. Apparently they do not take the type traits check for integrality into account. I will the support for floating point values from this function as this can be achieved without ambiguity using `std::fabs`, making the intended use to be able to take the abs of any integral type number without compiler warnings.

**Comment by wlambooy** [comment]
Windows CI build is out of heap space again 🙄

File: include/fiction/utils/math_utils.hpp [hunk 8]
```diff
@@ -27,6 +29,26 @@ T round_to_n_decimal_places(const T number, const uint64_t n)
...
 
+/**
+ * Takes the absolute value of an integral number if it is signed, and otherwise computes the identity. This avoids a
+ * compiler warning when taking the absolute value of an unsigned number.
+ * @tparam T The type of the number to take the absolute value of. Must be integral.
+ * @param n The number to take the absolute value of.
+ * @return |n|.
+ */
+template <typename T>
+T integral_abs(const T n)
```
Comment (marcelwa) [inline include/fiction/utils/math_utils.hpp:40]:
```suggestion
T integral_abs(const T n) noexcept
```

File: include/fiction/utils/math_utils.hpp [hunk 9]
```diff
@@ -7,17 +7,19 @@
...
- * @param number the number to round.
- * @param n the number of decimal places to round to.
- * @return the number rounded to n decimal places.
+ * @tparam T The type of the number to round.
+ * @param number The number to round.
+ * @param n The number of decimal places to round to.
+ * @return The number rounded to n decimal places.
  */
 template <typename T>
 T round_to_n_decimal_places(const T number, const uint64_t n)
```
Comment (marcelwa) [inline include/fiction/utils/math_utils.hpp:25]: I know you didn't touch this one, but could you add `noexcept` for consistency?

```suggestion
T round_to_n_decimal_places(const T number, const uint64_t n) noexcept
```

**Review by marcelwa — CHANGES_REQUESTED** [review]
Looks essentially done. I just spotted a small inconsistency. After this, it's done, I promise :laughing:

**Review by marcelwa — APPROVED** [review]
Many thanks!
